### PR TITLE
Config improvements: optional database storage using SettingsStore

### PR DESCRIPTION
**Pull request by brusch**
Initial PR for #9330 and #6987

### CODE REVIEW

File: lib/Db/PimcoreConfigStorage.php [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    protected function load()
+    {
+        // load settingsstore data
+        $settingsStoreData = [];
+        $scope = $this->buildSettingsStoreScope();
+        $settingsStoreDataIds = SettingsStore::getIdsByScope($scope);
+        foreach ($settingsStoreDataIds as $settingsStoreDataId) {
+            $settingsStoreEntry = SettingsStore::get($settingsStoreDataId, $scope);
+            $settingsStoreEntryData =
+            $settingsStoreEntryData = json_decode($settingsStoreEntry->getData(), true);
```
Comment (NiklasBr) [inline lib/Db/PimcoreConfigStorage.php]:
```suggestion
            $settingsStoreEntryData = json_decode($settingsStoreEntry->getData(), true);
```

File: bundles/CoreBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -1249,6 +1251,70 @@ private function addTargetingNode(ArrayNodeDefinition $rootNode)
             ->end();
     }
 
+    private function addConfigurationSectionsNode(ArrayNodeDefinition $rootNode)
```
Comment (brusch) [inline bundles/CoreBundle/DependencyInjection/Configuration.php]: We don't need that anymore, we're just using the standard Symfony Config approach.

File: bundles/CoreBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -1249,6 +1251,70 @@ private function addTargetingNode(ArrayNodeDefinition $rootNode)
...
+                    ->end()
+                ->end()
+            ->end();
+    }
+
+    private function addImageThumbnailsNode(ArrayNodeDefinition $rootNode)
+    {
+        $rootNode
+            ->children()
+                ->arrayNode('imagethumbnails')
```
Comment (brusch) [inline bundles/CoreBundle/DependencyInjection/Configuration.php]: This should be in this hierarchy: 
```yml
pimcore: 
  asset: 
    thumbnails: 
      definitions: 
        ....
```

File: lib/Kernel.php [hunk 4]
```diff
@@ -155,6 +155,11 @@ public function registerContainerConfiguration(LoaderInterface $loader)
         if (file_exists($systemConfigFile)) {
             $loader->load($systemConfigFile);
         }
+
+        if (is_dir(PIMCORE_CONFIGURATION_DIRECTORY . '/imagethumbnails/')) {
```
Comment (brusch) [inline lib/Kernel.php]:
```suggestion
        if (is_dir(PIMCORE_CONFIGURATION_DIRECTORY . '/image-thumbnails/')) {
```
We're using `-` in other places as well - just to be consistent 😊

File: models/Asset.php [hunk 5]
```diff
@@ -537,18 +536,15 @@ public function save()
                     // if the old path is different from the new path, update all children
                     $updatedChildren = [];
                     if ($oldPath && $oldPath != $this->getRealFullPath()) {
-                        $differentOldPath = $oldPath;
-
                         try {
                             $storage->move($oldPath, $this->getRealFullPath());
+                            $differentOldPath = $oldPath;
```
Comment (brusch) [inline models/Asset.php]: How are all the changes in `models/Asset.php` related to this PR? Maybe a merging issue? This changes were introduced in #9693

File: models/Asset/Image/Thumbnail/Config.php [hunk 6]
```diff
@@ -20,8 +20,6 @@
 use Pimcore\Tool\Serialize;
 
 /**
- * @method \Pimcore\Model\Asset\Image\Thumbnail\Config\Dao getDao()
```
Comment (brusch) [inline models/Asset/Image/Thumbnail/Config.php:23]: Why removed that? Still valid right?

File: lib/Model/Dao/PimcoreConfigBag.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+namespace Pimcore\Model\Dao;
+
+use Pimcore\Db\PimcoreConfigStorage;
+
+/**
+ * @internal
+ */
+abstract class PimcoreConfigBag implements DaoInterface
```
Comment (brusch) [inline lib/Model/Dao/PimcoreConfigBag.php]:
```suggestion
abstract class PimcoreConfigBagDao implements DaoInterface
```

File: lib/Model/Dao/PimcoreConfigBag.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configure()
+    {
+        // nothing to do
+    }
+
+    /**
```
Comment (brusch) [inline lib/Model/Dao/PimcoreConfigBag.php]: How does saving work? 
Here you're calling `save()` on this dao class, but actually there's no: 
https://github.com/pimcore/pimcore/pull/9882/files#diff-de2fb2d6c9873638e6632e713bcbbf15099a44862159b86493ded6662d693b84R959

File: lib/Db/PimcoreConfigStorage.php [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+use Pimcore\Config;
+use Pimcore\Model\Tool\SettingsStore;
+use Sabre\VObject\Settings;
+use Symfony\Component\Yaml\Yaml;
+
+/**
+ * @internal
+ */
+final class PimcoreConfigStorage
```
Comment (brusch) [inline lib/Db/PimcoreConfigStorage.php]: I'd move all this code directly into the `PimcoreConfigBagDao` , I don't think we need to put that into an extra class.

File: lib/Model/Dao/PimcoreConfigBag.php (hunk 8 above)
Comment (weisswurstkanone) [reply lib/Model/Dao/PimcoreConfigBag.php]: image

Comment (weisswurstkanone) [reply lib/Model/Dao/PimcoreConfigBag.php]: image

File: models/Asset/Image/Thumbnail/Config.php (hunk 6 above)
Comment (weisswurstkanone) [reply models/Asset/Image/Thumbnail/Config.php:23]: because Pimcore\Model\Asset\Image\Thumbnail\Config now has its own implementation of save. So the annotation would be invalid.

File: lib/Model/Dao/PimcoreConfigBag.php (hunk 7 above)
Comment (weisswurstkanone) [reply lib/Model/Dao/PimcoreConfigBag.php]: changed in a different commit

File: models/Asset.php [hunk 10]
```diff
@@ -1553,7 +1553,7 @@ public function setCustomSettings($customSettings)
     /**
      * @return string
      */
-    public function getMimeType()
+    public function getMimetype()
```
Comment (brusch) [inline models/Asset.php]: @weisswurstkanone here are still some changes which do not belong to this PR.

File: lib/Db/PimcoreConfigStorage.php (hunk 9 above)
Comment (brusch) [reply lib/Db/PimcoreConfigStorage.php]: PimcoreConfigBagListingDao for listing

File: bundles/CoreBundle/DependencyInjection/Configuration.php [hunk 11]
```diff
@@ -457,6 +457,48 @@ private function addAssetNode(ArrayNodeDefinition $rootNode)
...
+                                                ->scalarNode('name')->end()
+                                                ->scalarNode('description')->end()
+                                                ->scalarNode('group')->end()
+                                                ->scalarNode('format')->end()
+                                                ->scalarNode('quality')->end()
+                                                ->scalarNode('highResolution')->end()
+                                                ->booleanNode('preserveColor')->end()
+                                                ->booleanNode('preserveMetaData')->end()
+                                                ->scalarNode('rasterizeSVG')->end()
+                                                ->scalarNode('downloadable')->end()
```
Comment (dvesh3) [inline bundles/CoreBundle/DependencyInjection/Configuration.php]:
```suggestion
                                                ->booleanNode('rasterizeSVG')->end()
                                                ->booleanNode('downloadable')->end()
```

File: lib/Model/Dao/PimcoreConfigBagDao.php [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+     * @deprecated Will be removed in Pimcore 11
+     */
+    private static ?PhpArrayFileTable $legacyStore = null;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configure()
+    {
+        $params = func_get_arg(0);
```
Comment (dvesh3) [inline lib/Model/Dao/PimcoreConfigBagDao.php:61]:
```suggestion
    public function configure(...$params)
    {
```

File: models/Asset/Image/Thumbnail/Config/Listing/Dao.php [hunk 13]
```diff
@@ -15,49 +15,40 @@
...
+use Pimcore\Model\Tool\SettingsStore;
 
 /**
  * @internal
  *
  * @property \Pimcore\Model\Asset\Image\Thumbnail\Config\Listing $model
+ * @property PimcoreConfigStorage $db
  */
-class Dao extends Model\Dao\PhpArrayTable
+class Dao extends Config\Dao
```
Comment (dvesh3) [inline models/Asset/Image/Thumbnail/Config/Listing/Dao.php:29]:
```suggestion
class Dao extends Model\Dao\PimcoreConfigBagDao
```

File: models/Asset/Image/Thumbnail/Config/Listing/Dao.php [hunk 14]
```diff
@@ -15,49 +15,40 @@
...
+use Pimcore\Db\PimcoreConfigStorage;
 use Pimcore\Model;
 use Pimcore\Model\Asset\Image\Thumbnail\Config;
+use Pimcore\Model\Tool\SettingsStore;
 
 /**
  * @internal
  *
  * @property \Pimcore\Model\Asset\Image\Thumbnail\Config\Listing $model
+ * @property PimcoreConfigStorage $db
```
Comment (dvesh3) [inline models/Asset/Image/Thumbnail/Config/Listing/Dao.php:27]:
```suggestion
```

File: models/Asset/Image/Thumbnail/Config/Listing/Dao.php [hunk 15]
```diff
@@ -15,49 +15,40 @@
 
 namespace Pimcore\Model\Asset\Image\Thumbnail\Config\Listing;
 
+use Pimcore\Db\PimcoreConfigStorage;
```
Comment (dvesh3) [inline models/Asset/Image/Thumbnail/Config/Listing/Dao.php]:
```suggestion
```

File: models/Asset/Image/Thumbnail/Config/Listing/Dao.php (hunk 14 above)
Comment (dvesh3) [inline models/Asset/Image/Thumbnail/Config/Listing/Dao.php]:
```suggestion
```

File: lib/Model/Dao/PimcoreConfigBagDao.php [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+        if ($this->dataSource && !file_exists($yamlFilename) && $this->dataSource !== self::DATA_SOURCE_LEGACY) {
+            // this configuration already exists so check if it is writeable
+            // this is only the case if it comes from var/config or from the legacy file, or the settings-store
+            // however, we never want to write it back to the legacy file
+
+            throw new \Exception(sprintf('Configuration can only be written to %s, however the config comes from a different source', $yamlFilename));
+        }
+
+        File::put($yamlFilename, Yaml::dump($data, 5));
```
Comment (dvesh3) [inline lib/Model/Dao/PimcoreConfigBagDao.php]:
```suggestion
        File::put($yamlFilename, Yaml::dump($data, 50));
```
Otherwise, the file content will look like
```yaml
pimcore:
    assets:
        image:
            thumbnails:
                definitions: { Test: { items: [{ method: contain, arguments: { width: 100, height: 100, forceResize: false } }], medias: {  }, name: Test, description: '', group: '', format: SOURCE, quality: 85, highResolution: !!float 0, preserveColor: false, preserveMetaData: false, rasterizeSVG: false, downloadable: false, modificationDate: 1628597508, creationDate: 1628597487, preserveAnimation: false } }

```

File: lib/Model/Dao/PimcoreConfigBagDao.php [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+    /**
+     * @param string $id
+     * @param array $data
+     * @throws \Exception
+     */
+    protected function saveData(string $id, $data)
+    {
+        $writeLocation = $this->getWriteTarget();
+
+        if ($writeLocation === "yaml") {
```
Comment (NiklasBr) [inline lib/Model/Dao/PimcoreConfigBagDao.php]:
```suggestion
        if ($writeLocation === self::WRITE_TARGET_YAML) {
```

File: lib/Model/Dao/PimcoreConfigBagDao.php [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+            $this->writeYaml($id, $data);
+        } else if ($writeLocation === "settings-store") {
```
Comment (NiklasBr) [inline lib/Model/Dao/PimcoreConfigBagDao.php]:
```suggestion
        } else if ($writeLocation === self::WRITE_TARGET_SETTINGS_STORE) {
```

**Comment by brusch** [comment]
@NiklasBr thanks a lot! 👍
